Mathlib/ModelTheory/PartialEquiv.lean

```diff
@@ -269,6 +273,189 @@ theorem toPartialEquiv_toEmbedding {f :  M ≃ₚ[L] N} (h : f.dom = ⊤) :
+          toFun := (fun i ↦ (S i).dom)
+          monotone' := monotone_dom.comp S.monotone}).symm)
+
+@[simp]
+theorem dom_partialEquivLimit : (partialEquivLimit S).dom = iSup (fun x ↦ (S x).dom) := rfl
+
+@[simp]
+theorem cod_partialEquivLimit : (partialEquivLimit S).cod = iSup (fun x ↦ (S x).cod) := rfl
+
+@[simp]
+lemma partialEquivLimit_comp_inclusion {i : ι} :
+    (partialEquivLimit S).toEquiv.toEmbedding.comp (Substructure.inclusion (le_iSup _ i)) =
+    (Substructure.inclusion (le_iSup _ i)).comp (S i).toEquiv.toEmbedding := by
+  simp only [partialEquivLimit, Equiv.comp_toEmbedding, Embedding.comp_assoc]
+  rw [Equiv_isup_symm_inclusion]
+  congr
+
+theorem le_partialEquivLimit : ∀ i, S i ≤ partialEquivLimit S :=
+  fun i => ⟨le_iSup (f := fun i ↦ (S i).dom) _, by
+    simp only [cod_partialEquivLimit, dom_partialEquivLimit, partialEquivLimit_comp_inclusion, ←
+      Embedding.comp_assoc, subtype_comp_inclusion]⟩
+
+end DirectLimit
+
+section FGEquiv
+
+open PartialEquiv Set DirectLimit
+
+variable (M) (N) (L)
+
+/-- The type of equivalences between finitely generated substructures. -/
+abbrev FGEquiv := {f : M ≃ₚ[L] N // f.dom.FG}
+
+/-- Two structures `M` and `N` form an extension pair if the domain of any finitely-generated map
+from `M` to `N` can be extended to include any element of `M`. -/
+def IsExtensionPair : Prop := ∀ (f : L.FGEquiv M N) (m : M), ∃ g, m ∈ g.1.dom ∧ f ≤ g
+
+variable {M N L}
+
+theorem countable_self_fgequiv_of_countable [Countable M] :
+    Countable (L.FGEquiv M M) := by
+  let g : L.FGEquiv M M →
+      Σ U : { S : L.Substructure M // S.FG }, U.val →[L] M :=
+    fun f ↦ ⟨⟨f.val.dom, f.prop⟩, (subtype _).toHom.comp f.val.toEquiv.toHom⟩
+  have g_inj : Function.Injective g := by
+    intro f f' h
+    ext
+    let ⟨⟨dom_f, cod_f, equiv_f⟩, f_fin⟩ := f
+    cases congr_arg (·.1) h
+    apply PartialEquiv.ext (by rfl)
+    simp only [g, Sigma.mk.inj_iff, heq_eq_eq, true_and] at h
+    exact fun x hx ↦ congr_fun (congr_arg (↑) h) ⟨x, hx⟩
+  have : ∀ U : { S : L.Substructure M // S.FG }, Structure.FG L U.val :=
+    fun U ↦ (U.val.fg_iff_structure_fg.1 U.prop)
+  exact Function.Embedding.countable ⟨g, g_inj⟩
+
+instance inhabited_self_FGEquiv : Inhabited (L.FGEquiv M M) :=
+  ⟨⟨⟨⊥, ⊥, Equiv.refl L (⊥ : L.Substructure M)⟩, fg_bot⟩⟩
+
+/-- Maps to the symmetric finitely-generated partial equivalence. -/
+def FGEquiv.symm (f : L.FGEquiv M N) : L.FGEquiv N M := ⟨f.1.symm, f.1.dom_fg_iff_cod_fg.1 f.2⟩
+
+lemma IsExtensionPair_iff_cod : L.IsExtensionPair M N ↔
+    ∀ (f : L.FGEquiv N M) (m : M), ∃ g, m ∈ g.1.cod ∧ f ≤ g := by
+  refine Iff.intro ?_ ?_ <;>
+  · intro h f m
+    obtain ⟨g, h1, h2⟩ := h f.symm m
+    exact ⟨g.symm, h1, monotone_symm h2⟩
+
+namespace IsExtensionPair
+
+protected alias ⟨cod, _⟩ := IsExtensionPair_iff_cod
+
+/-- The cofinal set of finite equivalences with a given element in their domain. -/
+def definedAtLeft
+    (h : L.IsExtensionPair M N) (m : M) : Order.Cofinal (FGEquiv L M N) where
```

Comment:
I don't have a better name to suggest, but I should point out that this name reads like a Prop. Maybe you have an idea to improve it? Non-blocking, though

Reply:
Would `cofinalMemDom` or `memDomCofinal` be better?
(It might read like a Prop one way or another, because cofinality is a subtype that easily could have been a Prop)

Reply:
I would have guessed that it would be a predicate to be cofinal, but it seems to be implemented as a subtype of sets in analogy to ideals, where it does make sense to be bundled. I have no particularly strong opinion, as I'm not really aware of the other API features.

Reply:
`Order.Ideal` is a half-baked file so I am willing to hear that the API is suboptimal. I will try writing you a `Final`/`Cofinal` API tomorrow.

Reply:
To be clear - I don't have any real complaint, or even a clear request, with the `Order.Cofinal` API, in part because I didn't write any of the proofs that really use it here, @QuinnLesquimau did.

Reply:
I guess this can be future work